playset/session.py

```diff
@@ -0,0 +1,214 @@
+        self.dataDir = str(cog_data_path(self))
+        if not 'cards' in self.dataDir:
+            url = 'https://frky-storage.s3-us-west-1.amazonaws.com/cards.zip'
+            wget.download(url, self.dataDir+'/cards.zip')
+            zip = ZipFile(self.dataDir+'/cards.zip')
+            zip.extractall(self.dataDir)
+            zip.close()
+        self.dataDir = self.dataDir+'/cards'
+        self.ctx = ctx
+        self.scores = Counter()
+        self.deck = np.random.permutation(81)
+        self.boardW = 4
+        self.board = np.zeros((3,self.boardW),dtype=int)
+        for i in range(self.board.size):
+            self.board[i%3,i//3] = self.deck[0]
+            self.deck = self.deck[1:]
+        while not _board_contains_set(self.board):
+            self.boardW += 1
+            oldBoard = self.board
+            self.board = np.zeros((3,self.boardW),dtype=int)
+            for i in range(oldBoard.size):
+                self.board[i%3][i//3] = oldBoard[i%3][i//3]
+            for i in range(3):
+                self.board[i][self.boardW-1] = self.deck[0]
+                self.deck = self.deck[1:]
+        self._gen_board_image()
+        
+    @classmethod
+    def start(cls, ctx):
+        session = cls(ctx)
+        loop = ctx.bot.loop
+        session._task = loop.create_task(session.run())
+        return session
+        
+    async def run(self):
+        await self._send_startup_msg()
+        self.game_running = True
+        while self.game_running:
+            await asyncio.sleep(2)
+            f = discord.File(self.dataDir+'/board.png')
+            await self.ctx.send(file=f)
+            foundSet = await self.wait_for_set()
+            await self._update_board(foundSet)
+            if _board_contains_set(self.board):
+                self._gen_board_image()
+            else:
+                self.game_running = False
+                
+        await self.end_game()
+        
+    async def _send_startup_msg(self):
+        await self.ctx.send("Starting Set. Type in the three card letters to call a set. Incorrect calls are -1 point. Good luck")
+        await asyncio.sleep(3)
+
+    async def wait_for_set(self):
+        self.foundSet = False
+        self.wrongAnswers = []
+
+        message = await asyncio.gather(self.ctx.bot.wait_for("message", check=self.check_set),self._wrong_handler())
+        guess = message[0].content.upper()
+        cards = []
+        for i in range(3):
+            cards.append(self.board[_LETTER_MAP[guess[i]]])
+        self.scores[message[0].author] += 1
+        await self.ctx.send(str(message[0].author)+": Set! +1 point")
+   
+        return cards
+        
+    async def _wrong_handler(self):
+        while ((not self.foundSet) or len(self.wrongAnswers)>0):
+            if(len(self.wrongAnswers)>0):
+                m = self.wrongAnswers[0]
+                self.scores[m.author] -= 1
+                await self.ctx.send(str(m.author)+": not a set. -1 point")
```

Comment:
For tracking, author is what is wanted, since that is linked to their true name (at least that is what trivia does). For display, there is author.display_name which can be used.

Reply:
I agree it's correct for tracking, I'm just specifically concerned about this message to the channel behaving weirdly. `display_name` would be a reasonable adjustment